### PR TITLE
Add detection for Recursion in Java

### PR DESCRIPTION
**Pull request by DarkaMaul**
To prepare for the blogpost publication.

### CODE REVIEW

File: java/src/docs/security/Recursion/Recursion.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    // ...
+}
+```
+In this example, a binary stream reader processes tokens recursively.
+
+For each new token \`0x2\`, the parser will create a new recursive call. If this stream is user-controlled, an attacker can generate too many stackframes and crash the application with a `StackOverflow` error.
+
+
+## References
+* Trail Of Bits Blog: [Low-effort denial of service with recursion](https://blog.trailofbits.com/2024/05/16/TODO/)
```
Comment (mschwager) [inline java/src/docs/security/Recursion/Recursion.md]: TODO

**Review by mschwager — COMMENTED** [review]
Overall, this LGTM. I added @GrosQuildu for review too.

This isn't a blocker, but I wonder if we can generalize the `RecursiveCallOrderN` functionality. It'd be nice to be able to specify arbitrary depth, and have the default value be 4. I think that would end up creating a cleaner query too, but probably requires some more advanced CodeQL functionality. Anyway, I was just curious if you considered this.

Also, have you done any [MVRA](https://docs.github.com/en/code-security/codeql-for-vs-code/getting-started-with-codeql-for-vs-code/running-codeql-queries-at-scale-with-multi-repository-variant-analysis) runs to check for false positives and such?

File: java/src/security/Recursion/Recursion.qhelp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    <example>
+        <sample src="RecursiveCall.java" />
+        <p>In this example, a binary stream reader processes tokens recursively.</p>
+        <p>For each new token `0x2`, the parser will create a new recursive call.
+        If this stream is user-controlled, an attacker can generate too many stackframes
+        and crash the application with a <code>StackOverflow</code> error.</p>
+    </example>
+    <references>
+        <li>
+            Trail Of Bits Blog: <a href="https://blog.trailofbits.com/2024/05/16/TODO/">Low-effort denial of service with recursion</a>
```
Comment (GrosQuildu) [inline java/src/security/Recursion/Recursion.qhelp]: TODO

File: java/src/security/Recursion/Recursion.ql [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+import java
+
+predicate isTestPackage(RefType referenceType) {
+  referenceType.getPackage().getName().toLowerCase().matches("%test%") or
+  referenceType.getPackage().getName().toLowerCase().matches("%benchmark%") or
+  referenceType.getName().toLowerCase().matches("%test%")
+}
+
+predicate isBefore(Method a, Method b) {
```
Comment (GrosQuildu) [inline java/src/security/Recursion/Recursion.ql]: Using lines of code for analysis is susp. Try to use CFG (https://codeql.github.com/codeql-standard-libraries/java/semmle/code/java/controlflow/BasicBlocks.qll/predicate.BasicBlocks$BasicBlock$bbDominates.1.html) or call graph (https://codeql.github.com/docs/codeql-language-guides/navigating-the-call-graph/).

**Comment by GrosQuildu** [comment]
I added more tests in one commit (please keep them).

In the second one, I proposed a much-simplified query that finds any recursive loops. From MVRA the query is speedy enough, but returns a lot of false positives. Would be great to limit FPs before publication with at least this heuristic: do not report if there is a depth/level argument (any number) that is changed for any call in the recursion call path.

Moreover, would be nice to print the offensive method and all methods in the call-path, so debugging is easier. We would need to use path-query instead of the problem one. Then we could also trim results by showing only one method from a cycle (e.g., for `level1->level2->level3` now we show 3 results).

Added these TODOs in query code.

Lastly, finding unbounded recursive paths with high precision is probably a hard problem. Would be nice to figure additional heuristics to limit false positives. Probably the "call depth depends on user-controlled input" is the useful one, and we can model it in codeql relatively easily.

**Review by GrosQuildu — DISMISSED** [review]

File: java/src/security/Recursion/Recursion.ql (hunk 3 above)
Comment (GrosQuildu) [reply java/src/security/Recursion/Recursion.ql]: Ah, it is for results deduplication. Interesting

File: java/test/query-tests/security/Recursion/Recursion.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+class RecursiveCallLimited {
+    // ok: recursion is limited
+    public boolean directRecursiveDepth(int depth) {
+        if (depth == 0) {
+            return true;
+        }
+        return directRecursiveDepth(depth - 1);
+    }
+
+    // todook: recursion is limited
```
Comment (DarkaMaul) [inline java/test/query-tests/security/Recursion/Recursion.java:161]: I could not make this one not be flagged :(

File: java/src/security/Recursion/Recursion.ql [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  override string toString() {
+    result = this.getCaller().toString() + " calls " + this.getCallee().toString()
+  }
+}
+
+/**
+ * Check if the Expr uses directly an argument of the enclosing function
+ */
+class ParameterOperation extends Expr {
+  ParameterOperation() {
```
Comment (GrosQuildu) [inline java/src/security/Recursion/Recursion.ql:33]: We may need parentheses here; now it reads:
```
(this instanceof BinaryExpr)
or
(
this instanceof UnaryAssignExpr
and
exists(...)
)
```

File: java/src/security/Recursion/Recursion.ql [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    exists(VarAccess va | va.getVariable() = this.getEnclosingCallable().getAParameter() |
+      this.getAChildExpr+() = va
+    )
+  }
+}
+
+module RecursiveConfig implements DataFlow::StateConfigSig {
+  class FlowState = Method;
+
+  predicate isSource(DataFlow::Node node, FlowState state) {
```
Comment (GrosQuildu) [inline java/src/security/Recursion/Recursion.ql]: maybe lets call `state` something like `firstCall`? should be easier to read

File: java/test/query-tests/security/Recursion/Recursion.java (hunk 4 above)
Comment (GrosQuildu) [reply java/test/query-tests/security/Recursion/Recursion.java:161]: May need a local data flow; but that's not a very often pattern I guess

File: java/src/security/Recursion/Recursion.ql (hunk 5 above)
Comment (DarkaMaul) [reply java/src/security/Recursion/Recursion.ql:33]: Great catch!

**Comment by DarkaMaul** [comment]
I updated the query to solve massive performance issues (it was timing out on most repositories): 

1. The source is now more restricted : instead of trying to store MethodCall, we store `Method` and already check that they are part of recursions chains (this restricted sources from 600k results to 500 on Elasticsearch)
2. Consider `getCallee` instead of `getCaller` to further reduce the number of candidates 
3. 😢 Remove the deduplication logic: I have tried several approaches, but they all failed for large codebase. I don't know if it's a bug with `isBarrierIn/Out` with `FlowState`, or something I don't understand about how the predicate is resolved.

**Comment by DarkaMaul** [comment]
Of note, the CI is failing for a Go query :/

**Comment by GrosQuildu** [comment]
LGTM; we can solve deduplication later. Please open an issue for it and please describe the unsuccessful attempts.

From curiosity, could you check what is the performance and false-positive rate of https://github.com/trailofbits/codeql-queries/pull/14/commits/c3213e4a2ee66766a163fc214a861f47650d334b version?

CI is failing for another query, gonna fix it in a new PR.

**Review by mschwager — APPROVED** [review]
Approving to move forward with external references.

**Review by fegge — APPROVED** [review]

**Pull request merged**